File: Assignment-08.java

// Problem Statement: Implement shell sort algorithm in JAVA demonstrating 
// shell data structure with modularity of programming language.


PROGRAM >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

  
import java.util.Scanner;

class ShellSort {
    
    // Function to perform Shell Sort
    public static void shellSort(int[] arr) {
        int n = arr.length;
        
        // Start with a large gap and reduce it
        for (int gap = n / 2; gap > 0; gap /= 2) {
            
            // Perform gapped insertion sort for this gap size
            for (int i = gap; i < n; i++) {
                int temp = arr[i];
                int j = i;
                
                // Shift elements to find correct position for arr[i]
                while (j >= gap && arr[j - gap] > temp) {
                    arr[j] = arr[j - gap];
                    j -= gap;
                }
                
                // Place temp at its correct position
                arr[j] = temp;
            }
        }
    }

    // Function to display the array
    public static void displayArray(int[] arr) {
        for (int num : arr) {
            System.out.print(num + " ");
        }
        System.out.println();
    }

    // Main function
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Taking user input
        System.out.print("Enter number of elements: ");
        int n = scanner.nextInt();
        
        int[] arr = new int[n];
        
        System.out.println("Enter elements of the array:");
        for (int i = 0; i < n; i++) {
            arr[i] = scanner.nextInt();
        }

        System.out.println("\nOriginal Array:");
        displayArray(arr);
        
        // Sorting the array using Shell Sort
        shellSort(arr);

        System.out.println("\nSorted Array using Shell Sort:");
        displayArray(arr);
        
        scanner.close();
    }
}

END OF CODE >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
  

File: Assignment-08.py

// Problem Statement: Implement shell sort algorithm in JAVA demonstrating 
// shell data structure with modularity of programming language.


PROGRAM >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

  
def shell_sort(arr):
    n = len(arr)
    gap = n // 2

    # Start with a large gap, then reduce the gap
    while gap > 0:
        # Perform gapped insertion sort
        for i in range(gap, n):
            temp = arr[i]
            j = i

            # Shift earlier gap-sorted elements up until the correct location for arr[i] is found
            while j >= gap and arr[j - gap] > temp:
                arr[j] = arr[j - gap]
                j -= gap

            arr[j] = temp
        gap //= 2

def display_array(arr):
    print(" ".join(map(str, arr)))

def main():
    # Taking user input
    n = int(input("Enter number of elements: "))
    arr = []

    print("Enter elements of the array:")
    for _ in range(n):
        arr.append(int(input()))

    print("\nOriginal Array:")
    display_array(arr)

    # Sorting the array using Shell Sort
    shell_sort(arr)

    print("\nSorted Array using Shell Sort:")
    display_array(arr)

if __name__ == "__main__":
    main()

END OF CODE >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
  
